src/libcrypto/poly_math.c

```diff
@@ -1,39 +1,254 @@
 #include "include_asm.h"
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", null);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_null);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_add);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", copy);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_copy);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", single_to_double);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", double_to_single);
-
-INCLUDE_ASM("asm/non_matchings/libcrypto/poly_math", multiply_shift);
+#include "libcrypto/bb_nn.h"
+#include "bbtypes.h"
+#include "macros.h"
+
+void null(field_2n* a) {
+    element* eptr = a->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        *eptr++ = 0;
+    }
+}
+
+void double_null(field_double* a) {
+    element* eptr = a->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        *eptr++ = 0;
+    }
+}
+
+void double_add(field_double* a, field_double* b, field_double* c) {
+    element* aptr = a->e;
+    element* bptr = b->e;
+    element* cptr = c->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++) {
+        int tempA = *aptr++;
+        int tempB = *bptr++;
+        *cptr++ = tempA ^ tempB;
+    }
+}
+
+void copy(field_2n* from, field_2n* to) {
+    element* fromptr = from->e;
+    element* toptr = to->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(from->e); i++) {
+        *toptr++ = *fromptr++;
+    }
+}
+
+void double_copy(field_double* from, field_double* to) {
+    element* fromptr = from->e;
+    element* toptr = to->e;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(from->e); i++) {
+        *toptr++ = *fromptr++;
+    }
+}
+
+void single_to_double(field_2n* from, field_double* to) {
+    short i;
+
+    double_null(to);
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(from->e); i++) {
+        to->e[i + ARRAY_COUNT(to->e) - ARRAY_COUNT(from->e)] = from->e[i];
+    }
+}
+
+void double_to_single(field_double* from, field_2n* to) {
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(to->e); i++) {
+        to->e[i] = from->e[i + ARRAY_COUNT(from->e) - ARRAY_COUNT(to->e)];
+    }
+}
+
+void multiply_shift(field_double* a) {
+    element* eptr = &a->e[ARRAY_COUNT(a->e) - 1];
+    element bit = 0;
+    short i;
+
+    for (i = 0; i < (unsigned)ARRAY_COUNT(a->e); i++, eptr--) {
+        element temp = *eptr;
+        *eptr = temp * 2 | bit;
```

Comment:
```suggestion
        *eptr = (temp << 1) | bit;
```
I think this makes it a bit clearer what the intention is?